src/hooks/useLogbook.tsx

```diff
@@ -0,0 +1,475 @@
+              draft: {
+                ...state.logbooks[action.payload.tokenId]?.draft,
+                ...action.payload.draft,
+              },
+            },
+          },
+        }
+        break
+      default:
+        throw new Error()
+    }
+
+    stateRef.current = newState
+    return newState
+  }
+
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  /**
+   * Get logbook
+   */
+  const normalizedLogs = (logs: any[]): Log[] =>
+    logs.map(log => ({
+      sender: log.sender,
+      message: log.message,
+      createdAt: new Date(
+        (log.createdAt as ethers.BigNumber).toNumber() * 1000
+      ),
+      etherscan: toEtherscanUrl(""), // TODO: get tx hash from event
+    }))
+
+  // retrieve single logbook by a given tokenId
+  const getLogbook = async (tokenId: string) => {
+    // mark as loading
+    dispatch({
+      type: "updateLogbook",
+      payload: {
+        tokenId,
+        logbook: {
+          ...state.logbooks[tokenId],
+          loading: true,
+          error: "",
+          tokenId,
+        },
+      },
+    })
+
+    try {
+      const contract = new ethers.Contract(
+        env.contractAddress,
+        env.contractABI,
+        // ethers.getDefaultProvider(env.supportedChainId, { infura: env.infuraId })
+        new ethers.providers.InfuraProvider(env.supportedChainId, env.infuraId)
+      )
+
+      // retrieve lobook and token owner from contract
+      const logbook = await contract.readLogbook(tokenId)
+      const tokenOwner = await contract.ownerOf(tokenId)
+
+      dispatch({
+        type: "updateLogbook",
+        payload: {
+          tokenId,
+          logbook: {
+            loading: false,
+            error: "",
+
+            tokenId,
+            tokenOwner,
+
+            isLocked: logbook.isLocked,
+            logs: normalizedLogs(logbook.logs),
+
+            openSeaURL: toOpenSeaNFTUrl(tokenId),
+          },
+        },
+      })
+    } catch (err) {
+      console.error(err)
+      const errorMsg = getWalletErrorMessage({ error: err as Error, lang })
+
+      dispatch({
+        type: "updateLogbook",
+        payload: {
+          tokenId,
+          logbook: {
+            ...state.logbooks[tokenId],
+            loading: false,
+            error: errorMsg,
+            tokenId,
+          },
+        },
+      })
+    }
+  }
+
+  // retrieve all logbooks own by current account
+  const getOwnNFTs = async () => {
+    if (!account) {
+      return
+    }
+
+    // mark as loading
+    dispatch({
+      type: "update",
+      payload: {
+        ownNFTs: { ...state.ownNFTs, loading: true, error: "" },
+      },
+    })
+
+    try {
+      const contract = new ethers.Contract(
+        env.contractAddress,
+        env.contractABI,
+        // ethers.getDefaultProvider(env.supportedChainId, { infura: env.infuraId })
+        new ethers.providers.InfuraProvider(env.supportedChainId, env.infuraId)
+      )
+
```

Comment:
another suggestion: can we call the no cost `contract.balanceOf(address)` here before opensea apis? and only call it when `balance > 0` ; may cut some number of opensea calls to avoid rate limit

Reply:
do we allow anonymous visitors to view logbooks?

Reply:
> and only call it when balance > 0 ; 

Could we make this check on the components that need this business logic using `useAccount` hook?